domain/port/state/state.go

```diff
@@ -0,0 +1,271 @@
+	"github.com/juju/collections/set"
+	"github.com/juju/collections/transform"
+	"github.com/juju/errors"
+
+	"github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	internaldatabase "github.com/juju/juju/internal/database"
+	"github.com/juju/juju/internal/uuid"
+)
+
+// State represents the persistence layer for opened ports.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new state reference.
+func NewState(factory database.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// GetOpenedPorts returns the opened ports for a given unit,
+// grouped by endpoint.
+func (st *State) GetOpenedPorts(ctx context.Context, unitUUID string) (network.GroupedPortRanges, error) {
+	db, err := st.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	m := sqlair.M{
+		"unitUUID": unitUUID,
+	}
+
+	getOpenedPorts, err := st.Prepare(`
+SELECT &endpointPortRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+JOIN unit_endpoint ON port_range.unit_endpoint_uuid = unit_endpoint.uuid
+WHERE unit_endpoint.unit_uuid = $M.unitUUID
+`, endpointPortRange{}, m)
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing get opened ports statement")
+	}
+
+	results := []endpointPortRange{}
+	err = db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, getOpenedPorts, m).GetAll(&results)
+		if errors.Is(err, sqlair.ErrNoRows) {
+			return nil
+		}
+		return domain.CoerceError(err)
+	})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	groupedPortRanges := network.GroupedPortRanges{}
+	for _, endpointPortRange := range results {
+		if _, ok := groupedPortRanges[endpointPortRange.Endpoint]; !ok {
+			groupedPortRanges[endpointPortRange.Endpoint] = []network.PortRange{}
+		}
+		groupedPortRanges[endpointPortRange.Endpoint] = append(groupedPortRanges[endpointPortRange.Endpoint], decodeEndpointPortRange(endpointPortRange))
+	}
+	return groupedPortRanges, nil
+}
+
+// getOpenedPorts returns the opened ports for a given unit's endpoint.
+func (st *State) getOpenedPorts(ctx context.Context, tx *sqlair.TX, unitEndpointUUID unitEndpointUUID) ([]portRange, error) {
+	var result []portRange
+
+	selectOpenedPorts, err := st.Prepare(`
+SELECT &portRange.*
+FROM port_range
+JOIN protocol ON port_range.protocol_id = protocol.id
+WHERE unit_endpoint_uuid = $unitEndpointUUID.uuid
+`, portRange{}, unitEndpointUUID)
+
+	if err != nil {
+		return nil, errors.Annotate(err, "preparing select opened ports statement")
+	}
+
+	err = tx.Query(ctx, selectOpenedPorts, unitEndpointUUID).GetAll(&result)
+	if errors.Is(err, sqlair.ErrNoRows) {
+		return []portRange{}, nil
+	}
+	return result, domain.CoerceError(err)
+}
+
+// UpdateUnitPorts opens and closes ports for a given unit. Ports are grouped by endpoint.
+//
+// NOTE: This method opens ports first, and then closes ports. However, it is recommended
```

Comment:
Make this a hard requirement (in the comment) rather than a recommendation.

Reply:
Or just reconcile here.

Reply:
Done